### PR TITLE
Apply coverage and plateau models

### PR DESCRIPTION
**Pull request by averywpx**

### CODE REVIEW

File: gnomad_constraint/utils/constraint.py [hunk 1]
```diff
@@ -204,3 +207,231 @@ def create_constraint_training_dataset(
...
+        additional_grouping=grouping,
+        partition_hint=partition_hint,
+        use_table_group_by=True,
+    )
+    possible_ht = annotate_with_mu(possible_ht, mutation_ht)
+    possible_ht = possible_ht.transmute(possible_variants=possible_ht.variant_count)
+
+    # Apply plateau models and coverage model.
+    possible_ht = annotate_mutation_type(
+        possible_ht.annotate(mu_agg=possible_ht.mu_snp * possible_ht.possible_variants)
```
Comment (averywpx) [inline gnomad_constraint/utils/constraint.py:359]: I will think of a better name for it. The x of plateau model was `mu_snp` and here is `possible_ht.mu_snp * possible_ht.possible_variants`. I may need to ask Konrad what it is. Please let me know if you have any better idea!

File: gnomad_constraint/pipeline/constraint_pipeline.py [hunk 2]
```diff
@@ -175,6 +195,70 @@ def main(args):
...
+            for region in GENOMIC_REGIONS:
+                apply_models(
+                    preprocess_resources[(region, "exomes")].ht(),
+                    preprocess_resources[(region, "context")].ht(),
+                    mutation_ht,
+                    hl.experimental.read_expression(models[(region, "plateau")]),
+                    hl.experimental.read_expression(models[(region, "coverage")]),
+                    max_af=max_af,
+                    pops=POPS if use_pops else (),
+                    partition_hint=partition_hint * 20,
```
Comment (averywpx) [inline gnomad_constraint/pipeline/constraint_pipeline.py:255]: I'm not sure if it's the best way to do it. There are three different partition hint ([1](https://github.com/broadinstitute/gnomad-constraint/blob/3f4ebe90560cf4ddfb753b8dbe1d4e373a01d5eb/gnomad_constraint/utils/constraint.py#L114) [2](https://github.com/broadinstitute/gnomad-constraint/blob/3f4ebe90560cf4ddfb753b8dbe1d4e373a01d5eb/gnomad_constraint/utils/constraint.py#L226) [3](https://github.com/broadinstitute/gnomad-constraint/blob/3f4ebe90560cf4ddfb753b8dbe1d4e373a01d5eb/gnomad_constraint/utils/constraint.py#L425)). Should we made these separate parameters and arguments?

File: gnomad_constraint/utils/constraint.py [hunk 3]
```diff
@@ -196,11 +199,240 @@ def create_constraint_training_dataset(
...
+
+    # Count the possible variants in the context Table grouped by `keep_annotations`.
+    possible_ht = count_variants_by_group(
+        context_ht,
+        additional_grouping=grouping,
+        partition_hint=partition_hint,
+        use_table_group_by=True,
+    )
+    possible_ht = annotate_with_mu(possible_ht, mutation_ht)
+    possible_ht = possible_ht.transmute(possible_variants=possible_ht.variant_count)
```
Comment (jkgoodrich) [inline gnomad_constraint/utils/constraint.py:355]: Change to use the suggested function above.
```suggestion
    ht = create_observed_and_possible_ht(
        exome_ht,
        context_ht,
        mutation_ht,
        max_af,
        keep_annotations,
        pops,
        grouping,
        partition_hint,
        filter_coverage_over_0=True,
        filter_to_canonical_synonymous=False,
    )
```

File: gnomad_constraint/utils/constraint.py [hunk 4]
```diff
@@ -196,11 +199,240 @@ def create_constraint_training_dataset(
...
+    for pop in pops:
+        agg_expr[f"predicted_proportion_observed_{pop}"] = hl.agg.array_sum(
+            ht[f"predicted_proportion_observed_{pop}"]
+        )
+        agg_expr[f"expected_variants_{pop}"] = hl.agg.array_sum(
+            ht[f"expected_variants_{pop}"]
+        )
+        agg_expr[f"downsampling_counts_{pop}"] = hl.agg.array_sum(
+            ht[f"downsampling_counts_{pop}"]
+        )
```
Comment (jkgoodrich) [inline gnomad_constraint/utils/constraint.py:424]: I think this can be simplified by creating just aggregations without the annotations in between. If they are needed for computational reasons then it's fine to keep annotations, but can still probably be made less redundant and easier to read.

Here is the solution with removing the annotation steps. Obviously things need to be cleaned up, but I think this is enough to show how it can be done.

```suggestion
    mu_expr = ht.mu_snp * ht.possible_variants

    # Determine coverage correction to use based on coverage value
    cov_corr_expr = (
        hl.case()
          .when(ht.coverage == 0, 0)
          .when(ht.coverage >= cov_cutoff, 1)
          .default(coverage_model[1] * hl.log10(ht.coverage) + coverage_model[0])
    )

    # Generate sum aggregators for 'mu' on the entire dataset.
    agg_expr = {"mu": hl.agg.sum(mu_expr * cov_corr_expr)}
    agg_expr.update(name_me(ht, plateau_models, mu_expr, cov_corr_expr))

    for pop in pops:
        agg_expr.update(name_me(ht, plateau_models, mu_expr, cov_corr_expr, pop))

    # Aggregate the sum aggregators grouped by `grouping`
    grouping = list(grouping)
    grouping.remove("coverage")
```

Add this function:
```
def name_me(ht, plateau_models, mu_expr, cov_corr_expr, pop: Optional[str] = None):
    if pop is None:
        pop = ""
        plateau_model = hl.literal(plateau_models.total)[ht.cpg]
        slope = plateau_model[1]
        intercept = plateau_model[0]
        agg_func = hl.agg.sum
        ann_to_sum = ["observed_variants", "possible_variants"]
    else:
        plateau_model = hl.literal(plateau_models[pop])
        slope = hl.map(lambda f: f[ht.cpg][1], plateau_model)
        intercept = hl.map(lambda f: f[ht.cpg][0], plateau_model)
        agg_func = hl.agg.array_sum
        pop = f"_{pop}"
        ann_to_sum = [f"downsampling_counts{pop}"]

    # Apply plateau models for specified population.
    ppo_expr = mu_expr * slope + intercept

    # Generate sum aggregators for 'predicted_proportion_observed' and
    # 'expected_variants', for specified population.
    agg_expr = {
        f"predicted_proportion_observed{pop}": agg_func(ppo_expr),
        f"expected_variants{pop}": agg_func(ppo_expr * cov_corr_expr),
    }

    # Generate sum aggregators for 'observed_variants' and 'possible_variants' on
    # the entire dataset if pop is None, and for `downsampling_counts` for
    # specified population if pop is not None.
    agg_expr.update({ann: agg_func(ht[ann]) for ann in ann_to_sum})

    return agg_expr

```

**Review by jkgoodrich — CHANGES_REQUESTED** [review]
Some restructuring to remove redundancy, questions about naming, a couple bug fixes, and a suggestion on adding a checkpoint in a place that I think currently causes some portions to run twice. Looking a lot cleaner than the original, thanks for making it look more like the other similar function we worked on!

File: gnomad_constraint/pipeline/constraint_pipeline.py [hunk 5]
```diff
@@ -42,11 +42,13 @@
...
     get_sites_resource,
+    get_testing_dataset,
     get_training_dataset,
     mutation_rate_ht,
 )
 from gnomad_constraint.utils.constraint import (
     add_vep_context_annotations,
+    apply_models,
     create_constraint_training_dataset,
     prepare_ht_for_constraint_calculations,
```
[the remark this answers is not in the document]
Comment (averywpx) [reply gnomad_constraint/pipeline/constraint_pipeline.py:53]: I fixed it in the `build_model` branch as : 
```
context_lengths = list(filter(None, t.aggregate(hl.agg.collect_as_set(hl.len(t.context)))))
``` 
which has been merged to `main`.
This branch haven't been merged with the main branch yet so it wasn't been updated. I will merge and switch to your code!

File: gnomad_constraint/utils/constraint.py [hunk 6]
```diff
@@ -196,11 +199,240 @@ def create_constraint_training_dataset(
...
+            exome_ht, "worst_csq_by_gene"
+        )
+
+    else:
+        context_ht, _ = annotate_constraint_groupings(
+            context_ht, "transcript_consequences"
+        )
+        exome_ht, grouping = annotate_constraint_groupings(
+            exome_ht, "transcript_consequences"
+        )
```
[the remark this answers is not in the document]
Comment (averywpx) [reply gnomad_constraint/utils/constraint.py:311]: We may add other types of vep annotation later (see [here)](https://the-tgg.slack.com/archives/C03PP8Y98LA/p1658439544654799?thread_ts=1658438913.024999&cid=C03PP8Y98LA). Maybe use the string as for now will be more flexible? I will simplify it for sure!

File: gnomad_constraint/resources/resource_utils.py [hunk 7]
```diff
@@ -149,62 +156,95 @@ def get_models(
...
+        chr20, chrX, and chrY. Default is False.
+    :return: Path of the model.
+    """
+    check_param_scope(
+        version=version,
+        genomic_region=genomic_region,
+        custom_model=custom_model,
+    )
+    return TableResource(
+        f"{constraint_tmp_prefix}/{version}/model/applying/{custom_model}/constraint_applying.{genomic_region}{'.test' if test else ''}.ht"
```
[the remark this answers is not in the document]
Comment (averywpx) [reply gnomad_constraint/resources/resource_utils.py:183]: There is no other sub folder other than model. I will rearrange the path with your suggestions below!

File: gnomad_constraint/pipeline/constraint_pipeline.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkgoodrich) [reply gnomad_constraint/pipeline/constraint_pipeline.py:53]: Odd, in the main branch I see 
```
context_lengths = list(filter(None, set(hl.len(t.context).collect())))
```
https://github.com/broadinstitute/gnomad_methods/blob/c96646b0031bc3a057ed599490dad3d9f6b0d6c9/gnomad/utils/constraint.py#L314

[the remark this answers is not in the document]
Comment (averywpx) [reply gnomad_constraint/pipeline/constraint_pipeline.py:53]: So sorry this is the line I referred to. I copied the wrong line :(

File: gnomad_constraint/pipeline/constraint_pipeline.py [hunk 8]
```diff
@@ -214,6 +223,42 @@ def main(args):
...
+                overwrite,
+            )
+            # Apply coverage and plateau models for sites on autosomes/pseudoautosomal
+            # regions, chromosome X, and chromosome Y.
+            for region in GENOMIC_REGIONS:
+                apply_models(
+                    preprocess_resources[(region, "exomes")].ht(),
+                    preprocess_resources[(region, "context")].ht(),
+                    mutation_ht,
+                    hl.experimental.read_expression(models[(region, "plateau")]),
```
[the remark this answers is not in the document]
Comment (averywpx) [reply gnomad_constraint/pipeline/constraint_pipeline.py:251]: Thanks for the suggestion! I've add these to the files!

File: gnomad_constraint/utils/constraint.py [hunk 9]
```diff
@@ -196,11 +199,240 @@ def create_constraint_training_dataset(
...
+        gnomad_methods), formatted as a Tuple of intercept and slope, that calibrates a
+        given coverage level to observed:expected ratio. It's a correction factor for
+        low coverage sites.
+    :param max_af: Maximum allele frequency for a variant to be included in returned
+        counts. Default is 0.001.
+    :param keep_annotations: Annotations to keep in the context Table and exome Table.
+    :param pops: List of populations to use for downsampling counts. Default is ().
+    :param partition_hint: Target number of partitions for aggregation when counting
+        variants. Default is 2000.
+    :param custom_model: The customized model (one of "standard" or "worst_csq"),
```
[the remark this answers is not in the document]
Comment (averywpx) [reply gnomad_constraint/utils/constraint.py:286]: We may add other types of vep annotation later (see [here)](https://the-tgg.slack.com/archives/C03PP8Y98LA/p1658439544654799?thread_ts=1658438913.024999&cid=C03PP8Y98LA), so I felt that use the string as for now will be more flexible but I will change the parameter name. Possible models to be add can be found [here](https://github.com/broadinstitute/gnomad_lof/blob/e09d9fdc13d43c8ac996c93735ed206ba41ed473/constraint_utils/constraint_basics.py#L648).

File: gnomad_constraint/resources/resource_utils.py [hunk 10]
```diff
@@ -24,13 +25,30 @@
 DATA_TYPES = ["context", "exomes", "genomes"]
 MODEL_TYPES = ["plateau", "coverage"]
 GENOMIC_REGIONS = ["autosome_par", "chrx_nonpar", "chry_nonpar"]
+
+CUSTOM_VEP_ANNOTATIONS = ["transcript_consequences", "worst_csq_by_gene"]
```
Comment (averywpx) [inline gnomad_constraint/resources/resource_utils.py:29]: I replaced `cumstom_model` with `custom_vep_annotations` and added docstrings for this variable. My understanding is that the difference between them is they uses different vep annotations. Is this name accurate enough?

File: gnomad_constraint/pipeline/constraint_pipeline.py [hunk 11]
```diff
@@ -289,6 +346,36 @@ def main(args):
...
+        "--partition-hint-when-counting-variants",
+        help=(
+            "Target number of partitions for aggregation when counting variants and"
+            " applying models."
+        ),
+        type=int,
+        default=2000,
+    )
+    parser.add_argument(
+        "--partition-hint-for-sum-aggregators",
```
Comment (averywpx) [inline gnomad_constraint/pipeline/constraint_pipeline.py:364]: I felt that it's hard to make it more descriptive without making the name kind of long, but the variable is too wordy. Do you have any idea with the name?

File: gnomad_constraint/utils/constraint.py (hunk 4 above)
Comment (averywpx) [reply gnomad_constraint/utils/constraint.py]: I noticed that your code applied models on both observed and possible variants, while konrad's code only applied models to possible variants (see [here](https://github.com/broadinstitute/gnomad_lof/blob/e09d9fdc13d43c8ac996c93735ed206ba41ed473/constraint_utils/constraint_basics.py#L693)). Just for curiosity, was there a specific reason for this?

File: gnomad_constraint/utils/constraint.py [hunk 12]
```diff
@@ -195,12 +217,173 @@ def create_constraint_training_dataset(
     possible_ht = possible_ht.transmute(possible_variants=possible_ht.variant_count)
 
     # Outer join the Tables with possible variant counts and observed variant counts.
-    ht = ht.join(possible_ht, "outer")
-    # Annotate the Table with cpg site and mutation_type (one of "CpG", "non-CpG
+    ht = observed_ht.join(possible_ht, "outer")
+
+    tmp_path = new_temp_file(prefix="constraint", extension="ht")
+    ht.write(tmp_path)
+    ht = hl.read_table(tmp_path)
```
Comment (averywpx) [inline gnomad_constraint/utils/constraint.py:224]: Should I use hail's function `checkpoint` here or this is fine?

File: gnomad_constraint/utils/constraint.py [hunk 13]
```diff
@@ -196,11 +199,240 @@ def create_constraint_training_dataset(
...
+    possible_ht = annotate_mutation_type(
+        possible_ht.annotate(mu_agg=possible_ht.mu_snp * possible_ht.possible_variants)
+    )
+    # Apply plateau and coverage models for all sites.
+    total_plateau_model = hl.literal(plateau_models.total)[possible_ht.cpg]
+    ann_expr = {
+        "predicted_proportion_observed": possible_ht.mu_agg * total_plateau_model[1]
+        + total_plateau_model[0],
+        "coverage_correction": hl.case()
+        .when(possible_ht.coverage == 0, 0)
```
[the remark this answers is not in the document]
Comment (averywpx) [reply gnomad_constraint/utils/constraint.py]: I guess this line will remove the rows that are in `context_ht` but not `exome_ht` and with 0 coverage, these weren't filtered by `& (exome_ht.coverage > 0)`. I ran the code:
```
context_ht_v4_filtered = context_ht_v4.anti_join(exome_ht_v4.filter(keep_criteria, keep=False))
context_ht_v4_filtered.filter(context_ht_v4_filtered.exome_coverage == 0).count()
```
and the result was 5752294. Please correct me if I'm wrong.

File: gnomad_constraint/pipeline/constraint_pipeline.py [hunk 14]
```diff
@@ -289,6 +346,39 @@ def main(args):
...
+    parser.add_argument(
+        "--apply-models",
+        help=(
+            "Apply plateau and coverage models to variants in exome sites Table and"
+            " context Table to compute expected variant counts."
+        ),
+        action="store_true",
+    )
+    parser.add_argument(
+        "--partition-hint-when-counting-variants",
```
Comment (jkgoodrich) [inline gnomad_constraint/pipeline/constraint_pipeline.py:358]: Maybe:

```suggestion
        "--apply-obs-pos-count-partition-hint",
```

File: gnomad_constraint/pipeline/constraint_pipeline.py [hunk 15]
```diff
@@ -289,6 +346,39 @@ def main(args):
...
+        help=(
+            "Target number of partitions for aggregation when counting variants and"
+            " applying models."
```
Comment (jkgoodrich) [inline gnomad_constraint/pipeline/constraint_pipeline.py:361]:
```suggestion
            "Target number of partitions for aggregation when counting observed and expected variants for"
            " model application"
```

File: gnomad_constraint/pipeline/constraint_pipeline.py [hunk 16]
```diff
@@ -289,6 +346,39 @@ def main(args):
...
+        ),
+        type=int,
+        default=2000,
+    )
+    parser.add_argument(
+        "--partition-hint-for-sum-aggregators",
```
Comment (jkgoodrich) [inline gnomad_constraint/pipeline/constraint_pipeline.py:367]: Maybe:

```suggestion
        "--apply-expected-variant-partition-hint",
```

File: gnomad_constraint/pipeline/constraint_pipeline.py [hunk 17]
```diff
@@ -289,6 +346,39 @@ def main(args):
...
+        help="Target number of partitions for sum aggregators after applying models.",
```
Comment (jkgoodrich) [inline gnomad_constraint/pipeline/constraint_pipeline.py:368]:
```suggestion
        help="Target number of partitions for sum aggregators after applying models to get expected variant counts.",
```

File: gnomad_constraint/pipeline/constraint_pipeline.py [hunk 18]
```diff
@@ -66,9 +68,15 @@ def main(args):
     test = args.test
     overwrite = args.overwrite
     max_af = args.max_af
-    partition_hint = args.partition_hint
+    partition_hint_when_creating_training_set = (
+        args.partition_hint_when_creating_training_set
+    )
+    partition_hint_when_counting_variants = args.partition_hint_when_counting_variants
+    partition_hint_for_sum_aggregators = args.partition_hint_for_sum_aggregators
```
Comment (jkgoodrich) [inline gnomad_constraint/pipeline/constraint_pipeline.py:75]: Not changing these names since I don't know if you will agree with my suggested names

File: gnomad_constraint/pipeline/constraint_pipeline.py [hunk 19]
```diff
@@ -207,13 +221,53 @@ def main(args):
...
+                apply_models(
+                    preprocess_resources[(region, "exomes")].ht(),
+                    preprocess_resources[(region, "context")].ht(),
+                    mutation_ht,
+                    models[(region, "plateau")].he(),
+                    models[(region, "coverage")].he(),
+                    max_af=max_af,
+                    pops=POPS if use_pops else (),
+                    partition_hint_for_counting_variants=partition_hint_when_counting_variants,
+                    partition_hint_for_aggregation=partition_hint_for_sum_aggregators,
```
Comment (jkgoodrich) [inline gnomad_constraint/pipeline/constraint_pipeline.py:264]: Not changing names because I don't know if you will agree with my suggestions

File: gnomad_constraint/pipeline/constraint_pipeline.py [hunk 20]
```diff
@@ -289,6 +346,39 @@ def main(args):
...
+    parser.add_argument(
+        "--partition-hint-for-sum-aggregators",
+        help="Target number of partitions for sum aggregators after applying models.",
+        type=int,
+        default=1000,
+    )
+    parser.add_argument(
+        "--custom-vep-annotation",
+        help=(
+            "Custom which VEP annotation to be used to annotate transcript when"
```
Comment (jkgoodrich) [inline gnomad_constraint/pipeline/constraint_pipeline.py:375]:
```suggestion
            "Custom VEP annotation to be used to annotate transcript when"
```

File: gnomad_constraint/pipeline/constraint_pipeline.py [hunk 21]
```diff
@@ -289,6 +346,39 @@ def main(args):
...
+            ' applying models (one of "transcript_consequences" or "worst_csq_by_gene")'
+        ),
+        type=str,
+        default="transcript_consequences",
+        choices=["transcript_consequences", "worst_csq_by_gene"],
```
Comment (jkgoodrich) [inline gnomad_constraint/pipeline/constraint_pipeline.py:380]: Use `CUSTOM_VEP_ANNOTATIONS` from ressource_utils.py

File: gnomad_constraint/resources/resource_utils.py (hunk 10 above)
Comment (jkgoodrich) [reply gnomad_constraint/resources/resource_utils.py:29]: This looks good to me

File: gnomad_constraint/resources/resource_utils.py [hunk 22]
```diff
@@ -24,13 +25,30 @@
...
+"""
+VEP annotations used when applying models.
+
+"transcript_consequences" option will annotate the Table with 'annotation', 'gene', 'coverage', 'transcript', and 'canonical' annotations using 'transcript_consequences' VEP annotation.
+
+"worst_csq_by_gene" option will annotate the Table with 'annotation', 'gene', and 'coverage' annotations using 'worst_csq_by_gene' VEP annotation.
+"""
```
Comment (jkgoodrich) [inline gnomad_constraint/resources/resource_utils.py:36]: Looks good, just reformat length

File: gnomad_constraint/resources/resource_utils.py [hunk 23]
```diff
@@ -144,67 +162,92 @@ def get_models(
...
     )
-    return (
-        f"{constraint_tmp_prefix}/{version}/model/build_models/{model_type}.{genomic_region}{'.test' if test else ''}.he"
+    return ExpressionResource(
+        f"{constraint_tmp_prefix}/{version}/models/{model_type}.{genomic_region}{'.test' if test else ''}.he"
+    )
+
+
+def get_predicted_proportion_observed_dataset(
+    custom_vep_annotation: str,
```
Comment (jkgoodrich) [inline gnomad_constraint/resources/resource_utils.py:171]: Maybe make "transcript_consequences" the default

File: gnomad_constraint/resources/resource_utils.py [hunk 24]
```diff
@@ -144,67 +162,92 @@ def get_models(
...
+    version: str = CURRENT_VERSION,
+    genomic_region: str = "autosome_par",
+    test: bool = False,
+) -> str:
+    """
+    Return TableResource containing the expected variant counts and observed:expected ratio.
+
+    :param custom_vep_annotation: The customized model (one of "transcript_consequences" or "worst_csq_by_gene").
```
Comment (jkgoodrich) [inline gnomad_constraint/resources/resource_utils.py:179]:
```suggestion
    :param custom_vep_annotation: The VEP annotation used to customize the constraint model (one of "transcript_consequences" or "worst_csq_by_gene").
```

File: gnomad_constraint/resources/resource_utils.py [hunk 25]
```diff
@@ -214,6 +257,8 @@ def check_param_scope(
         "chrx_non_par", or "chry_non_par". Default is None.
     :param data_type: One of "exomes", "genomes" or "context". Default is None.
     :param model_type: One of "plateau", "coverage". Default is None.
+    :param custom_vep_annotation: The customized model (one of "transcript_consequences" or "worst_csq_by_gene").
```
Comment (jkgoodrich) [inline gnomad_constraint/resources/resource_utils.py:260]:
```suggestion
    :param custom_vep_annotation: The VEP annotation used to customize the constraint model (one of "transcript_consequences" or "worst_csq_by_gene").
```

File: gnomad_constraint/utils/constraint.py [hunk 26]
```diff
@@ -149,40 +157,54 @@ def create_constraint_training_dataset(
         `methylation_level` to group by when counting variants. Default is
         ('exome_coverage',).
     :param partition_hint: Target number of partitions for aggregation. Default is 100.
+    :param filter_coverage_over_0: Whether to filter the exome Table and context Table
+        to variants with coverage larger then 0. Default is False.
```
Comment (jkgoodrich) [inline gnomad_constraint/utils/constraint.py:161]:
```suggestion
        to variants with coverage larger than 0. Default is False.
```

File: gnomad_constraint/utils/constraint.py [hunk 27]
```diff
@@ -149,40 +157,54 @@ def create_constraint_training_dataset(
...
+    :param filter_to_canonical_synonymous: Whether to keep only canonical synonymous
+        variants in exome Table. Default is False.
```
Comment (jkgoodrich) [inline gnomad_constraint/utils/constraint.py:163]:
```suggestion
        variants in the exome Table. Default is False.
```

File: gnomad_constraint/utils/constraint.py (hunk 12 above)
Comment (jkgoodrich) [reply gnomad_constraint/utils/constraint.py:224]: I would use checkpoint, and then you don't need to save the name to "tmp_path" either

Comment (jkgoodrich) [reply gnomad_constraint/utils/constraint.py:224]: i would also add a tmp dir to your `hl.init(log="/constraint_pipeline.log")` in the pipeline script like here: https://github.com/broadinstitute/gnomad_qc/blob/08974a6d0a68e3456657f0db9b6b2bf6f356de47/gnomad_qc/v4/sample_qc/assign_ancestry.py#L373

File: gnomad_constraint/utils/constraint.py [hunk 28]
```diff
@@ -195,12 +217,173 @@ def create_constraint_training_dataset(
...
     # Outer join the Tables with possible variant counts and observed variant counts.
-    ht = ht.join(possible_ht, "outer")
-    # Annotate the Table with cpg site and mutation_type (one of "CpG", "non-CpG
+    ht = observed_ht.join(possible_ht, "outer")
+
+    tmp_path = new_temp_file(prefix="constraint", extension="ht")
+    ht.write(tmp_path)
+    ht = hl.read_table(tmp_path)
+
+    # Annotate the Table with cpg sites and mutation_type (one of "CpG", "non-CpG
```
Comment (jkgoodrich) [inline gnomad_constraint/utils/constraint.py:226]:
```suggestion
    # Annotate the Table with 'cpg' and 'mutation_type' (one of "CpG", "non-CpG
```

File: gnomad_constraint/utils/constraint.py [hunk 29]
```diff
@@ -149,40 +157,54 @@ def create_constraint_training_dataset(
         `methylation_level` to group by when counting variants. Default is
         ('exome_coverage',).
     :param partition_hint: Target number of partitions for aggregation. Default is 100.
+    :param filter_coverage_over_0: Whether to filter the exome Table and context Table
+        to variants with coverage larger then 0. Default is False.
+    :param filter_to_canonical_synonymous: Whether to keep only canonical synonymous
+        variants in exome Table. Default is False.
+    :param global_annotation: The annotation name of the StructExpression that contains
+        input parameters. Default is None.
```
Comment (jkgoodrich) [inline gnomad_constraint/utils/constraint.py:165]:
```suggestion
    :param global_annotation: The annotation name to use as a global StructExpression annotation containing
        input parameter values. If no value is supplied, this global annotation will not be added. Default is None.
```

File: gnomad_constraint/utils/constraint.py [hunk 30]
```diff
@@ -195,12 +217,173 @@ def create_constraint_training_dataset(
...
+        predicted_proportion_observed = sum(plateau model slop * mu_agg + plateau model intercept) (separately for CpG transitions and other sites)
+        if 0 < coverage < coverage cutoff:
+            coverage_correction = coverage_model slope * log10(coverage) + coverage_model intercept
+            expected_variants = sum(predicted_proportion_observed * coverage_correction)
+        else:
+            expected_variants = sum(predicted_proportion_observed)
+        The expected_variants are summed across the set of variants of interest to
+        obtain the final expected number of variants.
+
+    Function adds the following annotations:
```
Comment (jkgoodrich) [inline gnomad_constraint/utils/constraint.py:278]: Since all of these seem to be grouped by groupings you could just add that once here "... all grouped by groupings (output of `annotate_exploded_vep_for_constraint_groupings`)" and remove that specification below 


In other cases where you put a function name you also add `()` so maybe add that here too

File: gnomad_constraint/utils/constraint.py [hunk 31]
```diff
@@ -195,12 +217,173 @@ def create_constraint_training_dataset(
...
+
+    :param exome_ht: Exome sites Table (output of `prepare_ht_for_constraint_calculations
+        ()`) filtered to autosomes and pseudoautosomal regions.
+    :param context_ht: Context Table (output of `prepare_ht_for_constraint_calculations
+        ()`) filtered to autosomes and pseudoautosomal regions.
+    :param mutation_ht: Mutation rate Table with 'mu_snp' field.
+    :param plateau_models: Linear models (output of `build_models()` in
+        gnomad_methods`), with the values of the dictionary formatted as a
+        StrucExpression of intercept and slope, that calibrates mutation rate to
+        proportion observed for high coverage exome. It includes models for CpG s,
```
Comment (jkgoodrich) [inline gnomad_constraint/utils/constraint.py:301]:
```suggestion
        proportion observed for high coverage exome. It includes models for CpG sites,
```

File: gnomad_constraint/pipeline/constraint_pipeline.py (hunk 14 above)
Comment (averywpx) [reply gnomad_constraint/pipeline/constraint_pipeline.py:356]: I felt 'apply' may imply that it's only used in `apply_models`, so I change it to `count-obs-pos-vars-partition-hint`!

File: gnomad_constraint/utils/constraint.py (hunk 12 above)
Comment (averywpx) [reply gnomad_constraint/utils/constraint.py:224]: > I would use checkpoint, and then you don't need to save the name to "tmp_path" either

Since `new_temp_file` generates different paths every time, I didn't include `overwrite` parameter. Please let me know if it's the correct way to do.

Comment (averywpx) [reply gnomad_constraint/utils/constraint.py:224]: > i would also add a tmp dir to your `hl.init(log="/constraint_pipeline.log")` in the pipeline script like here: https://github.com/broadinstitute/gnomad_qc/blob/08974a6d0a68e3456657f0db9b6b2bf6f356de47/gnomad_qc/v4/sample_qc/assign_ancestry.py#L373

Do you mean the tmp dir (`constraint_tmp_prefix`) saves the intermediate files in pipeline? I used that instead of the temp dir generated by `new_temp_file`.

**Review by jkgoodrich — CHANGES_REQUESTED** [review]
A few small things I found when testing. 

Also, I think the reason there were different numbers is just because I forgot to add `--use-weights` to the run. Check the new file out: gs://gnomad-tmp/julia/constraint_new_testing/2.1.1/predicted_proportion_observed/transcript_consequences/predicted_proportion_observed.autosome_par.ht

**Comment by jkgoodrich** [comment]
I added a few small comments before you requested the review. I will wait until they are addressed before I look over the PR again

**Comment by averywpx** [comment]
> Also, I think the reason there were different numbers is just because I forgot to add `--use-weights` to the run. Check the new file out: gs://gnomad-tmp/julia/constraint_new_testing/2.1.1/predicted_proportion_observed/transcript_consequences/predicted_proportion_observed.autosome_par.ht

I double checked the result and they are identical. Thank you so much for debugging it!

File: gnomad_constraint/resources/resource_utils.py [hunk 32]
```diff
@@ -214,6 +260,8 @@ def check_param_scope(
         "chrx_non_par", or "chry_non_par". Default is None.
     :param data_type: One of "exomes", "genomes" or "context". Default is None.
     :param model_type: One of "plateau", "coverage". Default is None.
+    :param custom_vep_annotation: The VEP annotation used to customize the constraint model (one of "transcript_consequences" or "worst_csq_by_gene").
+        Default is None.
```
Comment (jkgoodrich) [inline gnomad_constraint/resources/resource_utils.py:264]:
```suggestion
        Default is None.
    :return: None.
```

**Review by jkgoodrich — COMMENTED** [review]
Very small suggested changes then it looks good to me

**Review by jkgoodrich — APPROVED** [review]
LGTM!
